tests/v2/test_1613-generator-tolayout-records.py

```diff
@@ -0,0 +1,46 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+import pytest  # noqa: F401
+import numpy as np  # noqa: F401
+import awkward as ak  # noqa: F401
+
+import awkward._v2._lookup  # noqa: E402
+import awkward._v2._connect.cling  # noqa: E402
+
+ROOT = pytest.importorskip("ROOT")
+
+
+compiler = ROOT.gInterpreter.Declare
+
+
+def test_RecordArray_NumpyArray():
+    array = ak._v2.contents.recordarray.RecordArray(
+        [
+            ak._v2.contents.numpyarray.NumpyArray(np.array([0, 1, 2, 3, 4], np.int64)),
+            ak._v2.contents.numpyarray.NumpyArray(
+                np.array([0.0, 1.1, 2.2, 3.3, 4.4, 5.5])
+            ),
+        ],
+        ["x", "y"],
+        parameters={"__record__": "Something"},
+    )
+    layout = array
+    generator = ak._v2._connect.cling.togenerator(layout.form, flatlist_as_rvec=False)
+    lookup = ak._v2._lookup.Lookup(layout)
+    generator.generate(compiler)
+
+    array_out = generator.tolayout(lookup, 0, ())
+    assert array.to_list() == array_out.to_list()
+    array_out = generator.tolayout(lookup, 0, ("x"))
```

Comment:
@jpivarski - do I assume correctly that we can either retrieve all of the fields or just one?

Reply:
First, be careful with replacing `()` (an empty tuple) with `("x")` (a single string: not a tuple!). It might work in this case because the `tolayout` code just iterates over what it has in the third argument, and that means _iterating over the letters of the string_, but this name is one letter long. Python's idea that strings are interchangeable with other iterables, by iterating over characters in the string, has been the source of many bugs. I think you wanted `("x",)` there.

But more generally for the question of what that third argument is for: it's useful in Numba but not C++. It allows field-lookups to be delayed so that

```python
array["x"][100]
```

can be interchangeable with

```python
array[100]["x"]
```

The third argument is a list of field names that are waiting to be evaluated (nested: `("x", "y", "z")` means _z_ inside _y_ inside _x_).

In C++, however, field-lookups don't have this "commutivity":

```c++
array.x()[100]
```

is not the same as

```c++
array[100].x()
```

So this third argument is dead-weight for C++. You shouldn't use it for anything cppyy or RDataFrame related, which is what tests/v2/test_1613-generator-tolayout-records.py is.

Reply:
Thanks! No, I do not use it anywhere except in the test to access the code branch to make sure it is covered.